docs/source/nested.rst

```diff
@@ -10,213 +10,471 @@ Introduction
-history from the tensors passed to the constructor. For more information, refer to the section on
-:ref:`constructor functions`.
+>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
+>>> print([component for component in nt])
+[tensor([0, 1, 2]), tensor([3, 4, 5, 6, 7])]
 
-In order to form a valid NestedTensor all the passed Tensors need to match in dimension, but none of the other attributes need to.
+Each tensor in the list must have the same number of dimensions, but the shapes can otherwise vary
+along a single dimension. If the dimensionalities of the input components don't match, the
+constructor throws an error.
 
->>> a = torch.randn(3, 50, 70) # image 1
->>> b = torch.randn(3, 128, 64) # image 2
->>> nt = torch.nested.nested_tensor([a, b], dtype=torch.float32)
->>> nt.dim()
-4
-
-If one of the dimensions doesn't match, the constructor throws an error.
-
->>> a = torch.randn(50, 128) # text 1
->>> b = torch.randn(3, 128, 64) # image 2
->>> nt = torch.nested.nested_tensor([a, b], dtype=torch.float32)
-Traceback (most recent call last):
-  File "<stdin>", line 1, in <module>
-RuntimeError: All Tensors given to nested_tensor must have the same dimension. Found dimension 3 for Tensor at index 1 and dimension 2 for Tensor at index 0.
+>>> a = torch.randn(50, 128) # 2D tensor
+>>> b = torch.randn(2, 50, 128) # 3D tensor
+>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
+...
+RuntimeError: When constructing a nested tensor, all tensors in list must have the same dim
 
-Note that the passed Tensors are being copied into a contiguous piece of memory. The resulting
-NestedTensor allocates new memory to store them and does not keep a reference.
+During construction, dtype, device, and whether gradients are required can be chosen via the
+usual keyword arguments.
 
-At this moment we only support one level of nesting, i.e. a simple, flat list of Tensors. In the future
-we can add support for multiple levels of nesting, such as a list that consists entirely of lists of Tensors.
-Note that for this extension it is important to maintain an even level of nesting across entries so that the resulting NestedTensor
-has a well defined dimension. If you have a need for this feature, please feel encouraged to open a feature request so that
-we can track it and plan accordingly.
+>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged, dtype=torch.float32, device="cuda", requires_grad=True)
+>>> print([component for component in nt])
+[tensor([0., 1., 2.], device='cuda:0',
+       grad_fn=<UnbindBackwardAutogradNestedTensor0>), tensor([3., 4., 5., 6., 7.], device='cuda:0',
+       grad_fn=<UnbindBackwardAutogradNestedTensor0>)]
 
-size
-+++++++++++++++++++++++++
-
-Even though a NestedTensor does not support ``.size()`` (or ``.shape``), it supports ``.size(i)`` if dimension i is regular.
+In the vein of ``torch.as_tensor``, ``torch.nested.as_nested_tensor`` can be used to preserve autograd
```

Comment:
This is preexisting, but I'm not sure I want to mention `torch.as_tensor` here. It seems to not actually preserve autograd history in the actual cases you would want to use it, e.g. convert from numpy.

Reply:
removed mention of it